File: src/python/core/evaluator.py
from core.parser import Parser
from core.symbols import SyntaxKind
from core.lexer import SyntaxToken
from core.expressions import NumberExpression, VariableExpression, UnaryExpression, BinaryExpression, CallExpression, PostfixExpression

from core.functions.functions import FUNCTION_INFO, BUILTIN_FUNCTION_STRING_MAP

from core.operators.factorial import factorial

from core.constants import PI, E
import math

POSTFIX_OPERATORS = {
    SyntaxKind.FACTORIAL_TOKEN: factorial,
    SyntaxKind.PERCENTAGE_TOKEN: lambda x: x * 0.01,
}

CONSTANTS = {
    "pi": PI,
    "e": E,
}

class Evaluator:
    def __init__(self, ast):
        self.ast = ast

    @classmethod
    def from_string(cls, input_string):
        parser = Parser.from_string(input_string)
        ast = parser.parse()
        return cls(ast)
    
    @classmethod
    def from_tokens(cls, tokens):
        parser = Parser.from_tokens(tokens)
        ast = parser.parse()
        return cls(ast)

    def evaluate(self):
        return self.ast.accept(self)

    def visit_number(self, node: NumberExpression):
        return node.value

    def visit_variable(self, node: VariableExpression):
        if node.name in CONSTANTS:
            return CONSTANTS[node.name]
        raise ValueError(f"Undefined variable: {node.name}")

    def visit_unary(self, node: UnaryExpression):
        operand = node.operand.accept(self)

        match node.operator.kind:
            case SyntaxKind.PLUS_TOKEN:
                return +operand
            case SyntaxKind.MINUS_TOKEN:
                return -operand
            case _:
                raise ValueError(f"Unsupported unary operator: {node.operator.kind}")

    def visit_binary(self, node: BinaryExpression):
        left = node.left.accept(self)
        right = node.right.accept(self)

        match node.operator.kind:
            case SyntaxKind.PLUS_TOKEN:
                return left + right
            case SyntaxKind.MINUS_TOKEN:
                return left - right
            case SyntaxKind.MULTIPLY_TOKEN:
                return left * right
            case SyntaxKind.DIVIDE_TOKEN:
                return left / right
            case _:
                raise ValueError(f"Unsupported binary operator: {node.operator.kind}")

    def visit_postfix(self, node: PostfixExpression):
        operand = node.operand.accept(self)

        if node.operator.kind in POSTFIX_OPERATORS:
            operator_function = POSTFIX_OPERATORS[node.operator.kind]
            return operator_function(operand)
        
        raise ValueError(f"Unsupported postfix operator: {node.operator.kind}")

    def visit_call(self, node: CallExpression):
        arguments = [node.accept(self) for node in node.arguments]
        # use inbuilt functions now
        match node.identifier:
            case "sin":
                return math.sin(arguments[0])
            case "cos":
                return math.cos(arguments[0])
            case "tan":
                return math.tan(arguments[0])
            case "pow":
                return math.pow(arguments[0], arguments[1])
            case "ln":
                return math.log(arguments[0])
            case "log":
                return math.log(arguments[0], 10)
            case "sqrt":
                return math.sqrt(arguments[0])
            case "nroot":
                return math.pow(arguments[0], 1/arguments[1])
            case "abs":
                return abs(arguments[0])
            case "frac":
                return arguments[0] / arguments[1]
        
        raise ValueError(f"Undefined function: {node.identifier}")

if __name__ == "__main__":
    input_string = "pow(2, pow(3, 2)) + nroot(27, 3) - ln(e) + 5!"
    evaluator = Evaluator(input_string)
    result = evaluator.evaluate()
    print(f"Result of evaluating expression: {input_string} is {result}")

File: src/c/core/evaluator.c
#include "evaluator.h"
#include "expression.h"
#include <math.h>
#include <string.h>
#include <stdio.h>

typedef enum {
    INBUILT_FUNCTION_INVALID = -1,
    INBUILT_FUNCTION_SIN,
    INBUILT_FUNCTION_COS,
    INBUILT_FUNCTION_TAN,
    INBUILT_FUNCTION_LN,
    INBUILT_FUNCTION_LOG,
    INBUILT_FUNCTION_SQRT,
    INBUILT_FUNCTION_POW,
    INBUILT_FUNCTION_ABS,
    INBUILT_FUNCTION_NROOT
} InbuiltFunction;

static InbuiltFunction get_inbuilt_function(const char* name) {
    if (strcmp(name, "sin") == 0) return INBUILT_FUNCTION_SIN;
    if (strcmp(name, "cos") == 0) return INBUILT_FUNCTION_COS;
    if (strcmp(name, "tan") == 0) return INBUILT_FUNCTION_TAN;
    if (strcmp(name, "ln") == 0) return INBUILT_FUNCTION_LN;
    if (strcmp(name, "log") == 0) return INBUILT_FUNCTION_LOG;
    if (strcmp(name, "sqrt") == 0) return INBUILT_FUNCTION_SQRT;
    if (strcmp(name, "pow") == 0) return INBUILT_FUNCTION_POW;
    if (strcmp(name, "abs") == 0) return INBUILT_FUNCTION_ABS;
    if (strcmp(name, "nroot") == 0) return INBUILT_FUNCTION_NROOT;

    return -1; // Invalid function
}

double call_inbuilt_function(InbuiltFunction func, double* args, size_t argc) {
    switch (func) {
        case INBUILT_FUNCTION_SIN:
            return sin(args[0]);
        case INBUILT_FUNCTION_COS:
            return cos(args[0]);
        case INBUILT_FUNCTION_TAN:
            return tan(args[0]);
        case INBUILT_FUNCTION_LN:
            return log(args[0]);
        case INBUILT_FUNCTION_LOG:
            return log10(args[0]);
        case INBUILT_FUNCTION_SQRT:
            return sqrt(args[0]);
        case INBUILT_FUNCTION_POW:
            return pow(args[0], args[1]);
        case INBUILT_FUNCTION_ABS:
            return fabs(args[0]);
        case INBUILT_FUNCTION_NROOT:
            if (args[1] == 0) {
                fprintf(stderr, "Error: Cannot take the 0th root\n");
                return NAN;
            }
            return pow(args[0], 1.0 / args[1]);
        default:
            fprintf(stderr, "Error: Unknown inbuilt function\n");
            return NAN;
    }
}

static double get_constant(const char* name) {
    if (strcmp(name, "pi") == 0) return M_PI;
    if (strcmp(name, "e") == 0) return M_E;

    return NAN;
}

double evaluate(const Expression* expressions, size_t index) {
    const Expression* expression = &expressions[index];

    switch (expression->type) {
        case EXPRESSION_TYPE_NUMBER:
            return expression->value.number;
        case EXPRESSION_TYPE_VARIABLE: {
            double constant_value = get_constant(expression->value.variable);
            if (!isnan(constant_value)) {
                return constant_value;
            } else {
                fprintf(stderr, "Error: Undefined variable '%s'\n", expression->value.variable);
                return NAN;
            }
        }
        case EXPRESSION_TYPE_UNARY: {
            double operand = evaluate(expressions, expression->value.unary.operand_index);
            switch (expression->value.unary.operator.kind) {
                case SYNTAX_KIND_PLUS:
                    return operand;
                case SYNTAX_KIND_MINUS:
                    return -operand;
                default:
                    fprintf(stderr, "Error: Unexpected unary operator %d\n", expression->value.unary.operator.kind);
                    return NAN;
            }
        }
        case EXPRESSION_TYPE_POSTFIX: {
            double operand = evaluate(expressions, expression->value.postfix.operand_index);
            switch (expression->value.postfix.operator.kind) {
                case SYNTAX_KIND_FACTORIAL: {
                    if (operand < 0 || floor(operand) != operand) {
                        fprintf(stderr, "Error: Factorial is only defined for non-negative integers, got %g\n", operand);
                        return NAN;
                    }
                    
                    if (operand > 170) { // 170! is the largest factorial that fits in a double
                        fprintf(stderr, "Error: Factorial result is too large to represent, got %g\n", operand);
                        return NAN;
                    }

                    double result = 1;
                    for (int i = 1; i <= (int)operand; i++) {
                        result *= i;
                    }

                    return result;
                }
                case SYNTAX_KIND_PERCENTAGE:
                    return operand / 100.0;
                default:
                    fprintf(stderr, "Error: Unexpected postfix operator %d\n", expression->value.postfix.operator.kind);
                    return NAN;
            }
        }
        case EXPRESSION_TYPE_BINARY: {
            double left = evaluate(expressions, expression->value.binary.left_index);
            double right = evaluate(expressions, expression->value.binary.right_index);
            switch (expression->value.binary.operator.kind) {
                case SYNTAX_KIND_PLUS:
                    return left + right;
                case SYNTAX_KIND_MINUS:
                    return left - right;
                case SYNTAX_KIND_MULTIPLY:
                    return left * right;
                case SYNTAX_KIND_DIVIDE:
                    if (right == 0) {
                        fprintf(stderr, "Error: Division by zero\n");
                        return NAN;
                    }
                    return left / right;
                default:
                    fprintf(stderr, "Error: Unexpected binary operator %d\n", expression->value.binary.operator.kind);
                    return NAN;
            }
        }
        case EXPRESSION_TYPE_CALL: {
            char* identifier = expression->value.call.identifier;
            InbuiltFunction func = get_inbuilt_function(identifier);

            if (func == INBUILT_FUNCTION_INVALID) {
                fprintf(stderr, "Error: Undefined function '%s'\n", identifier);
                return NAN;
            }

            double args[MAX_ARGUMENTS];
            for (size_t i = 0; i < expression->value.call.argument_count; i++) {
                args[i] = evaluate(expressions, expression->value.call.argument_indices[i]);
            }
            return call_inbuilt_function(func, args, expression->value.call.argument_count);
        }
    }
}